README.md
# music.dictionary

[![Build Status](https://travis-ci.org/danigb/tonal.svg?branch=master)](https://travis-ci.org/danigb/music.dictionary)
[![Test Coverage](https://codeclimate.com/github/danigb/music.dictionary/badges/coverage.svg)](https://codeclimate.com/github/danigb/music.dictionary/coverage)
[![Code Climate](https://codeclimate.com/github/danigb/music.dictionary/badges/gpa.svg)](https://codeclimate.com/github/danigb/music.dictionary)
[![js-standard-style](https://img.shields.io/badge/code%20style-standard-brightgreen.svg?style=flat)](https://github.com/feross/standard)
[![npm version](https://img.shields.io/npm/v/music.dictionary.svg)](https://www.npmjs.com/package/music.dictionary)
[![license](https://img.shields.io/npm/l/music.dictionary.svg)](https://www.npmjs.com/package/music.dictionary)
[![tonal](https://img.shields.io/badge/tonal-dictionary-yellow.svg)](https://www.npmjs.com/package/tonal)


`music.dictionary` is a function to create musical dictionaries. A musical dictionary is a collection of notes with a name:

```js
var dictionary = require('music.dictionary')
var chords = dictionary({'Maj7': ['1 3 5 7', ['maj7', 'M7']]})
chords('CMaj7') // => ['C', 'E', 'G', 'B']
chords('BbM7') // => ['Bb', 'D', 'F', 'A']
```

This is part of [tonal](https://www.npmjs.com/package/tonal) and the foundation of the [tonal.scale](https://github.com/danigb/tonal.scale) and [chord.dictionary](https://github.com/danigb/chord.dictionary) dictionaries.

## Install

Via npm only: `npm i --save music.dictionary`

## Usage

The function returned by `music.dictionary` creates list of notes (or intervals) from a name. The name may include a root:

```js
var chords = dictionary({'Maj7', ['1 3 5 7', ['maj7', 'M7']], 'm7': ['1 3m 7 7m']})
chords('CMaj7') // => ['C', 'E', 'G', 'B']
chords('Maj7', 'C') // => ['C', 'E', 'G', 'B']
chords('Fm7') // => ['F', 'Ab', 'C', 'Eb']
```

Without root, a list of intervals is returned:

```js
chords('Maj7') // => ['1P', '3M', '5P', '7M']
```

__It never returns null__. If no name is found, an empty array is returned:

```js
chords('Cdim') // => []
chords(null) // => []
chords() // => []
```

#### Dictionary data

The `data` property of the dictionary is a hashmap of names mapped to data objects. The data objects has the following properties:

- name: the name
- aliases: an array with the alternative names
- intervals: an array with the intervals
- steps: an array with the intervals in __array notation__
- binary: a binary representation of the set
- decimal: the decimal representation of the set

```js
var chords = dictionary({'Maj7', ['1 3 5 7', ['maj7', 'M7']]})
// dictionary data
chords.data['M7'] // => { name: 'Maj7', aliases: ['M7'],
                  //      intervals: ['1', '3', '5', '7'], steps: [ ...],
                  //      binary: '10010010001', decimal: 2193 }

// get chord by binary numbers
chords.data['100010010001'] === chords.data['Maj7']
chords.data[2193] === chords.data['Maj7']
```

#### Get available names

The property `names` is an array with the available names. The property `aliases` adds the aliases names:

```js
var chords = dictionary({'Maj7', ['1 3 5 7', ['maj7', 'M7']], 'm': ['1 3b 5']})
chords.names // => ['Maj7', 'm']
chords.aliases // => ['Maj7', 'm', 'maj7', 'M7']
```

#### More...

Read the [generated documentation](https://github.com/danigb/music.dictionary/blob/master/API.md) or take a look at [tonal](https://www.npmjs.com/package/tonal)

## License

MIT License


CHANGELOG
CHANGELOG

[0.0.5] Rename to music.dictionary. Remove detection code. (6.4kb / 5.1kb)
[0.0.4] Remove set dependency on dictionary build
[0.0.3] Update dependencies (10K)
[0.0.2] Get intervals with false as tonic (10K)
[0.0.1] First version (10K, 7.8K)


test/names.test.js
var vows = require('vows')
var assert = require('assert')
var dictionary = require('..')

var names = dictionary.names({'Maj7': ['1 3 5 7', ['maj7', 'M7']]}, ':')
vows.describe('dictionary.names').addBatch({
  'get names': function () {
    assert.deepEqual(names(), ['Maj7'])
    assert.deepEqual(names(false), ['Maj7'])
  },
  'get names with aliases': function () {
    assert.deepEqual(names(true), [ 'Maj7', 'maj7', 'M7' ])
  },
  'get name by notes': function () {
    assert.deepEqual(names('C E G B'), [ 'C:Maj7', 'C:maj7', 'C:M7' ])
  },
  'default separator is empty': function () {
    var names = dictionary.names({'m': ['1 3b 5']})
    assert.equal(names('F Ab C'), 'Fm')
  }
}).export(module)


test/getter.test.js
var vows = require('vows')
var assert = require('assert')
var dictionary = require('..')

var chords = dictionary.getter({'Maj7': ['1 3 5 7', ['maj7', 'M7']]})
vows.describe('dictionary.getter').addBatch({
  'create getter from data': function () {
    assert.deepEqual(dictionary.getter({'m': ['1 3b 5']})('Fm'), [ 'F', 'Ab', 'C' ])
  },
  'create a getter from a dictionary': function () {
    assert.deepEqual(dictionary.getter(dictionary({'m': ['1 3b 5']}))('Bm'), [ 'B', 'D', 'F#' ])
  },
  'get by type and tonic': function () {
    assert.deepEqual(chords('Maj7', 'C'), ['C', 'E', 'G', 'B'])
    assert.deepEqual(chords('M7', 'C'), ['C', 'E', 'G', 'B'])
    assert.deepEqual(chords('maj7', 'C'), ['C', 'E', 'G', 'B'])
  },
  'type with pitch class': function () {
    assert.deepEqual(chords('BbMaj7'), [ 'Bb', 'D', 'F', 'A' ])
    assert.deepEqual(chords('BbM7'), [ 'Bb', 'D', 'F', 'A' ])
    assert.deepEqual(chords('Bbmaj7'), [ 'Bb', 'D', 'F', 'A' ])
  },
  'type with note': function () {
    assert.deepEqual(chords('Eb3Maj7'), [ 'Eb3', 'G3', 'Bb3', 'D4' ])
    assert.deepEqual(chords('Eb3M7'), [ 'Eb3', 'G3', 'Bb3', 'D4' ])
    assert.deepEqual(chords('Eb3maj7'), [ 'Eb3', 'G3', 'Bb3', 'D4' ])
  },
  'get intervals': function () {
    assert.deepEqual(chords('Maj7', false), [ '1P', '3M', '5P', '7M' ])
    assert.deepEqual(chords('CMaj7', false), [ '1P', '3M', '5P', '7M' ])
  },
  'without tonic': function () {
    assert.deepEqual(chords('Maj7')('G'), [ 'G', 'B', 'D', 'F#' ])
  }
}).export(module)


test/dictionary.test.js
var vows = require('vows')
var assert = require('assert')
var dictionary = require('..')

vows.describe('music.dictionary').addBatch({
  'basic dictionary': {
    'topic': function () {
      return dictionary({'Maj7': ['1 3 5 7', ['maj7', 'M7']]})
    },
    'get by type and tonic': function (chords) {
      assert.deepEqual(chords('Maj7', 'C'), ['C', 'E', 'G', 'B'])
      assert.deepEqual(chords('M7', 'C'), ['C', 'E', 'G', 'B'])
      assert.deepEqual(chords('maj7', 'C'), ['C', 'E', 'G', 'B'])
    },
    'type with pitch class': function (chords) {
      assert.deepEqual(chords('BbMaj7'), [ 'Bb', 'D', 'F', 'A' ])
      assert.deepEqual(chords('BbM7'), [ 'Bb', 'D', 'F', 'A' ])
      assert.deepEqual(chords('Bbmaj7'), [ 'Bb', 'D', 'F', 'A' ])
    },
    'type with note': function (chords) {
      assert.deepEqual(chords('Eb3Maj7'), [ 'Eb3', 'G3', 'Bb3', 'D4' ])
      assert.deepEqual(chords('Eb3M7'), [ 'Eb3', 'G3', 'Bb3', 'D4' ])
      assert.deepEqual(chords('Eb3maj7'), [ 'Eb3', 'G3', 'Bb3', 'D4' ])
    },
    'get intervals': function (chords) {
      assert.deepEqual(chords('Maj7', false), [ '1P', '3M', '5P', '7M' ])
      assert.deepEqual(chords('CMaj7', false), [ '1P', '3M', '5P', '7M' ])
    },
    'without tonic': function (chords) {
      assert.deepEqual(chords('Maj7')('G'), [ 'G', 'B', 'D', 'F#' ])
    },
    'invalid values': function (chords) {
      assert.deepEqual(chords('Cblah'), [])
      assert.deepEqual(chords('blah'), [])
      assert.deepEqual(chords(), [])
    }
  }
}).export(module)


lib/names.js
'use strict'

var dictionary = require('./dictionary')
var binary = require('tonal.set/binary')
var split = require('tonal.gamut/split')

/**
 * Get a function to perform an inverse dictionary lookup (given notes, return names)
 *
 * @name dictionary.names
 * @function
 * @param {Array} names - the list of all names
 * @param {Hash} dictionary - the dictionary
 * @param {Function} builder - the name builder
 * @return {Function} a function to perform inverse lookup
 */
module.exports = function (d, sep) {
  var dict = typeof d === 'function' ? d : dictionary(d)
  sep = sep || ''
  return function (notes) {
    if (arguments.length === 0 || notes === false) return dict.names.slice()
    else if (notes === true) return dict.aliases.slice()
    var s = dict(binary(notes))
    if (!s) return []
    var tonic = split(notes)[0]
    return [ s.name ].concat(s.aliases).map(function (name) {
      return tonic + sep + name
    })
  }
}


lib/getter.js
'use strict'

var dictionary = require('./dictionary')
var transpose = require('tonal.transpose')
var split = require('tonal.notation/note.split')

/**
 * Given a dictionary return a function to get the notes or intervals from it
 *
 * @name dictionary.getter
 * @function
 * @param {Function|Hash} dictionary - the dictionary function or data
 * @return {Function} a function to get the notes or intervals from the set dictionary
 *
 * @example
 * var getter = require('tonal.dictionary/getter')
 * var get = getter({'Maj7': ['1 3 5 7'], 'm7': ['1 3b 5 7b'] })
 * get('CMaj7') // => ['C', 'E', 'G', 'B']
 */
module.exports = function (d) {
  var dict = typeof d === 'function' ? d : dictionary(d)
  return function get (name, tonic) {
    var v = dict(name)
    if (!v) {
      var n = split(name)
      v = n ? dict(n[5]) : null
      if (!v) return []
      tonic = tonic === false ? tonic : tonic || n[1] + n[2] + n[3]
    }
    if (tonic !== false && !tonic) return function (t) { return get(name, t) }
    return v.intervals.map(transpose(tonic))
  }
}


lib/dictionary.js
'use strict'

var parse = require('tonal.notation/interval.parse')

/**
 * Create a dictionary from a hash of names to intervals.
 *
 * The dictionary if a function that returns a data object from a given name.
 * The returned data object has the following properties:
 *
 * - name: the name
 * - aliases: an array with the alternative names
 * - intervals: an array with the intervals
 * - steps: an array with the intervals in __array notation__
 * - binary: a binary representation of the set
 * - decimal: the decimal representation of the set
 *
 * The dictionary itself has a `names` property with an array of available
 * names, and `aliases` property with an array of all names and aliases.
 *
 * The data source is a hash map of keys associated to intervals list, and optionally
 * an array of alternatives names:
 *
 * - A interval list (as string or as array)
 * - (Optionally) an array of name aliases
 *
 * See [chord.dictionary](https://github.com/danigb/chord.dictionary)
 *
 * @name dictionary
 * @function
 * @param {Hash} src - the data source
 * @return {Function} the dictionary
 *
 * @example
 * var dictionary = require('tonal.dictionary')
 * var chords = dictionary({'Maj7': ['1 3 5 7', ['maj7', 'M7']]})
 *
 * // get chord by name
 * chords('Maj7') // => { name: 'Maj7', aliases: ['maj7', 'M7'],
 *                //      intervals: ['1', '3', '5', '7'] }
 *
 * // get chord by aliases
 * chords('Maj7') === chords('maj7') === chords('M7')
 *
 * // get chord by binary numbers
 * chords('100010010001') === chords(22193) === chords('Maj7')
 *
 * // get chord names
 * chords.names // => ['Maj7']
 */
module.exports = function (src) {
  function dict (key) { return dict.data[key] }
  dict.source = src
  dict.names = Object.keys(src)
  dict.aliases = dict.names.slice()
  var data = dict.data = {}

  dict.names.forEach(function (k) {
    var d = src[k]
    var c = { name: k, aliases: d[1] || [] }
    c.intervals = d[0].split(' ')
    c.steps = c.intervals.map(parse)
    c.binary = binary([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], c.steps)
    c.decimal = parseInt(c.binary, 2)
    data[k] = data[c.binary] = data[c.decimal] = c
    c.aliases.forEach(function (a) { data[a] = c })
    if (c.aliases.length > 0) dict.aliases = dict.aliases.concat(c.aliases)
  })
  return dict
}

function binary (num, intervals) {
  intervals.forEach(function (i) { num[(i[0] * 7 + i[1] * 12) % 12] = '1' })
  return num.join('')
}


index.js
'use strict'

var parse = require('tonal.notation/interval.parse')
var split = require('tonal.notation/note.split')
var transpose = require('tonal.transpose')

/**
 * Create a musical dictionary. A musical dictionary is a function that given
 * a name (and optionally a tonic) returns an array of notes.
 *
 * A dictionary is created from a HashMap. It maps a name to a string with
 * an interval list and, optionally, an alternative name list (see example)
 *
 * Additionally, the dictionary has properties (see examples):
 *
 * - data: a hash with the dictionary data
 * - names: an array with all the names
 * - aliases: an array with all the names including aliases
 * - source: the source of the dictionary
 *
 * Each value of the data hash have the following properties:
 *
 * - name: the name
 * - aliases: an array with the alternative names
 * - intervals: an array with the intervals
 * - steps: an array with the intervals in __array notation__
 * - binary: a binary representation of the set
 * - decimal: the decimal representation of the set
 *
 * @name dictionary
 * @function
 * @param {Hash} source - the dictionary source
 * @return {Function} the dictionary
 *
 * @example
 * var dictionary = require('music.dictionary')
 * var chords = dictionary({'Maj7': ['1 3 5 7', ['M7']], 'm7': ['1 3b 5 7b'] })
 * chords('CMaj7') // => ['C', 'E', 'G', 'B']
 * chords('DM7') // => ['D', 'F#', 'A', 'C#']
 * chords('Bm7') // => ['B', 'D', 'F#', 'A']
 *
 * @example
 * // dictionary data
 * chords.data['M7'] // => { name: 'Maj7', aliases: ['M7'],
 *                   //      intervals: ['1', '3', '5', '7'], steps: [ ...],
 *                   //      binary: '10010010001', decimal: 2193 }
 *
 * // get chord by binary numbers
 * chords.data['100010010001'] === chords.data['Maj7']
 * chords.data[2193] === chords.data['Maj7']
 *
 * @example
 * // available names
 * chords.names // => ['Maj7', 'm7']
 * chords.aliases // => ['Maj7', 'm7', 'M7']
 */
module.exports = function (src) {
  function dict (name, tonic) {
    var v = dict.data[name]
    if (!v) {
      var n = split(name)
      v = n ? dict.data[n[5]] : null
      if (!v) return []
      tonic = tonic === false ? tonic : tonic || n[1] + n[2] + n[3]
    }
    if (tonic !== false && !tonic) return function (t) { return dict(name, t) }
    return v.intervals.map(transpose(tonic))
  }
  return build(src, dict)
}

function build (src, dict) {
  dict.source = src
  dict.names = Object.keys(src)
  dict.aliases = dict.names.slice()
  var data = dict.data = {}

  dict.names.forEach(function (k) {
    var d = src[k]
    var c = { name: k, aliases: d[1] || [] }
    c.intervals = d[0].split(' ')
    c.steps = c.intervals.map(parse)
    c.binary = binary([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], c.steps)
    c.decimal = parseInt(c.binary, 2)
    data[k] = data[c.binary] = data[c.decimal] = c
    c.aliases.forEach(function (a) { data[a] = c })
    if (c.aliases.length > 0) dict.aliases = dict.aliases.concat(c.aliases)
  })
  return dict
}

function binary (num, intervals) {
  intervals.forEach(function (i) { num[(i[0] * 7 + i[1] * 12) % 12] = '1' })
  return num.join('')
}
